working/bazel/module-structure.md

```diff
@@ -0,0 +1,234 @@
+
+In the Google mono-repository, there is a single root directory. Let's call it
+`google` - and it is commonly referenced by tools and Bazel with a prefix of
+`//`, so `//ninjacat` is logically located at `/google/ninjacat`. This root
+directory is called a _workspace_.
+
+Bazel [defines a _package_][2] as so:
+
+> The primary unit of code organization in a workspace is the _package_. A
+> package is a collection of related files and a specification of the
+> dependencies among them.
+>
+> A package is defined as a directory containing a file named `BUILD`,
+> residing beneath the top-level directory in the workspace. A package includes
+> all files in its directory, plus all subdirectories beneath it, except those
+> which themselves contain a BUILD file.
+
+[2]: https://docs.bazel.build/versions/master/build-ref.html#packages_targets
+
+As one can see, already _package_ is very different than a [`pub` package][3].
+
+[3]: https://www.dartlang.org/guides/libraries/create-library-packages
+
+We map the concept of `pub` packages with the following rules:
+
+* A `package:<a>.<b>/uri.dart` resolves to `//a/b/lib/uri.dart`.
+* Any other `package:<name>/uri.dart` resolves to `//third_party/dart/<name>`.
+
+### Inside a Package
+
+Once _inside_ a package, such as `//ninjacat/app`, you can expect the following:
+
+```
+> ninjacat/
+  > app/
+    > lib/
+    > test/
+    > BUILD
+```
+
+(If the package happens to be a Dart web _entrypoint_, you might also see `web/`
+and for VM binaries, you might also see `bin/`.)
+
+However, there is another important concept, [`targets`][4]:
+
+> A package is a container. The elements of a package are called _targets_. Most
+> targets are one of two principal kinds, files and rules. Additionally, there
+> is another kind of target, package groups, but they are far less numerous.
+
+[4]: https://docs.bazel.build/versions/master/build-ref.html#targets
+
+So, imagine the following, in the `BUILD` file:
+
+```
+dart_library(
+    name = "app",
+    srcs = ["lib/app.dart"],
+    deps = [
+         ":flags",
+    ],
+)
+
+dart_library(
+    name = "flags",
+    srcs = ["lib/flags.dart"],
+)
+
+dart_test(
+    name = "flags_test",
+    srcs = ["test/flags_test.dart"],
+    deps = [
+        ":flags",
+        "//third_party/dart/test",
+    ],
+)
+```
+
+Here we have _3_ targets:
+* `app`, which potentially is code that wraps together application-specific
+  code before being used later in the `main()` function of something in either
+  `web/` or `bin/`.
+* `flags`, which contains some common code for setting/getting flags.
+* `flags_test`, which tests that `flags` is working-as-intended.
+
+This concept already is very different than a `pub` package, where all of the
+files in `lib/` are sort of a single "implicit" target. In fact, a common issue
+externally is that `pubspec.yaml` (sort of similar to `BUILD`) is not granular
+enough, leading to the creation of "micro packages" that have a single file or
+capability.
```

Comment:
I tend to think of those micro packages as an anti-pattern. :( They can be useful if you *really* need to have different subsets of dependencies, but otherwise they just make maintenance and consumption of those packages harder.

Reply:
I think the most common cases, so far, which we haven't been able to help with:

* `{name}_test` packages (see below, basically avoiding a dependency on `package:test`)
* X-plat packages (i.e. a different implementation for web, I/O, flutter) might want different deps

Reply:
I understand the concerns here, but my impression is that fears around dependencies are somewhat overblown. Just depend on the union of the packages that the various platforms need. In almost all cases, the version solver should be able to handle them.

I know that hasn't always worked out well in practice, largely because the analyzer went through a period of very enthusiastic breaking version releases and many things transitively depend on it. But that's settled down and things should work pretty reliably now. Also, @nex3's new version solver is much better at handling these cases

Reply:
Now that a build system of some sort is essentially required for all Dart implementations, we should be a lot more careful about huge sets of extra dependencies.

The entire package graph and all the sources of all packages has to be considered during the build, and this can cause considerable bloat for things like DDC/Kernel Builders which operate on all files in all packages. Even if they never actually run because those files aren't imported, we still have to set up the asset graph such that they _could_ run on anything, and some initialization and finalizations steps do have to iterate all _potential_ nodes in the graph.